### PR TITLE
Monitor and protect distributor from OOMKilled due to too many in progress requests

### PR DESCRIPTION
**Pull request by alexqyle**
**What this PR does**:
Distributor could be overloaded if push requests to ingester getting slower. This would cause distributor uses more memory to hold timeseries objects from those slow requests and distributor could get OOMKilled when large requests kept piling up.

To protect distributor from this, 
- New config `max_inflight_push_requests` is introduced for ingester client. This is hard limit on how many request one distributor could send through one ingester client. Requests exceed this limit would be dropped.
- New metric `cortex_ingester_client_request_count` is created to keep monitoring how many requests were sent through each ingester client on each distributor. This could help to set `max_inflight_push_requests` with proper value.

**Which issue(s) this PR fixes**:
NA

**Checklist**
- [x] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Review by friedrichg — COMMENTED** [review]
How does this relate to ? or how is it better?
```
  # Max inflight push requests that this distributor can handle. This limit is
  # per-distributor, not per-tenant. Additional requests will be rejected. 0 =
  # unlimited.
  # CLI flag: -distributor.instance-limits.max-inflight-push-requests
  [max_inflight_push_requests: <int> | default = 0]
```

**Comment by alexqyle** [comment]
> How does this relate to ? or how is it better?
> 
> ```
>   # Max inflight push requests that this distributor can handle. This limit is
>   # per-distributor, not per-tenant. Additional requests will be rejected. 0 =
>   # unlimited.
>   # CLI flag: -distributor.instance-limits.max-inflight-push-requests
>   [max_inflight_push_requests: <int> | default = 0]
> ```

@friedrichg `max_inflight_push_requests` is controlling request coming into distributor. The new config is controlling request sending from distributor to ingester. With replication factor > 1 set, distributor would send time series to more than ingesters. 

For example, if RF set o 3, distributor would send same time series to 3 ingesters. According to DoBatch logic, distributor could accept new inflight request when 2 out of 3 requests are succeeded. However, if one of the ingester is unhealthy, the request to bad ingester would hanging there until timeout. At same time, distributor is accepting new inflight requests. In this case, inflight request to distributor does not change since 2 out of 3 ingesters returned 2xx, distributor would consider this inflight request completed. While request number from distributor to ingester increased, because one ingester is bad. Request to that bad ingester would take longer to complete. With new config introduced, number of request from distributor to that bad ingester could have a limit and make it fail fast.

Moreover, this new config could also prevent one distributor got overloaded if network connect is impaired between this distributor and ingesters.

File: pkg/ingester/client/client.go [hunk 1]
```diff
@@ -20,6 +22,19 @@ var ingesterClientRequestDuration = promauto.NewHistogramVec(prometheus.Histogra
 	Help:      "Time spent doing Ingester requests.",
 	Buckets:   prometheus.ExponentialBuckets(0.001, 4, 6),
 }, []string{"operation", "status_code"})
+var inflightRequestCount = promauto.NewGaugeVec(prometheus.GaugeOpts{
+	Namespace: "cortex",
+	Name:      "ingester_client_request_count",
```
Comment (friedrichg) [inline pkg/ingester/client/client.go:27]:
```suggestion
	Name:      "ingester_client_inflight_push_requests",
```
WDYT?. Definitely not ending in _count

**Review by friedrichg — CHANGES_REQUESTED** [review]
It's a great idea!. Thanks for explaining. I have just a minor nit

File: pkg/ingester/client/client.go (hunk 1 above)
Comment (alexqyle) [reply pkg/ingester/client/client.go:27]: Sounds good to me. I will make the change.

**Review by friedrichg — APPROVED** [review]
LGTM

**Comment by alanprot** [comment]
Are we cleaning the metric when we close the client ? Otherwise we will have dangling metrics during deployment or any other ingested replacement .. (as the metric is per ip)

**Comment by alexqyle** [comment]
> Are we cleaning the metric when we close the client ? Otherwise we will have dangling metrics during deployment or any other ingested replacement .. (as the metric is per ip)

Will create a new PR to address this

**Pull request merged**